### PR TITLE
=str #20550 only accept -1 for infinite retries in recoverWithRetries

### PR DESCRIPTION
**Pull request by samueltardieu**

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala [hunk 1]
```diff
@@ -803,11 +803,13 @@ final class Flow[-In, +Out, +Mat](delegate: scaladsl.Flow[In, Out, Mat]) extends
     * RecoverWithRetries allows to switch to alternative Source on flow failure. It will stay in effect after
     * a failure has been recovered up to `attempts` number of times so that each time there is a failure
     * it is fed into the `pf` and a new Source may be materialized. Note that if you pass in 0, this won't
-    * attempt to recover at all. Passing in a negative number will behave exactly the same as  `recoverWith`.
+    * attempt to recover at all. Passing in -1 will behave exactly the same as  `recoverWith`.
     *
     * Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements.
     * This stage can recover the failure signal, but not the skipped elements, which will be dropped.
     *
+    * `attempts` must be non-negative or equal to -1, otherwise IllegalArgumentException is thrown.
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/javadsl/Flow.scala]: `@param` for it (at bottom of scaladoc)?

File: akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowRecoverWithSpec.scala [hunk 2]
```diff
@@ -130,5 +130,14 @@ class FlowRecoverWithSpec extends AkkaSpec {
         .request(1)
         .expectError(ex)
     }
+
+    "throw IllegalArgumentException if number of retries is less than -1" in assertAllStagesStopped {
+      try {
+        Flow[Int].recoverWithRetries(-2, { case t: Throwable ⇒ Source.empty[Int] })
```
Comment (ktoso) [inline akka-stream-tests/src/test/scala/akka/stream/scaladsl/FlowRecoverWithSpec.scala]:
```scala
intercept[IllegalArgumentException] { 
 Flow[Int].recoverWithRetries(-2, { case t: Throwable ⇒ Source.empty[Int] })
}
```

**Comment by ktoso** [comment]
Thanks! LGTM after minor comments addressed

**Comment by ktoso** [comment]
Resolves #20550

**Comment by samueltardieu** [comment]
Comments addressed.

**Comment by johanandren** [comment]
LGTM

**Comment by ktoso** [comment]
LGTM, thanks!

**Pull request merged**